include/eve/module/core/regular/impl/simd/x86/gather.hpp

```diff
@@ -0,0 +1,162 @@
+    else if constexpr(i_64x4  && d_i32x4 )        return _mm256_i64gather_epi32(pi_t(p),v,4);
+    else if constexpr(i_64x2  && d_i32x4 )        return _mm_i64gather_epi32   (pi_t(p),v,4);
+    // 32 bits float data
+    else if constexpr(i_32x16 && c == float32x16) return _mm512_i32gather_ps(v, p, 4);
+    else if constexpr(i_32x8  && c == float32x8 ) return _mm256_i32gather_ps(p, v, 4);
+    else if constexpr(i_32x4  && c == float32x4 ) return _mm_i32gather_ps   (p, v, 4);
+    else if constexpr(i_64x8  && c == float32x8 ) return _mm512_i64gather_ps(v, p, 4);
+    else if constexpr(i_64x4  && c == float32x4 ) return _mm256_i64gather_ps(p, v, 4);
+    else if constexpr(i_64x2  && c == float32x4 ) return _mm_i64gather_ps   (p, v, 4);
+    // Aggregation cases
+    else if constexpr(has_aggregated_abi_v<wide<U, N>>)
+    {
+      auto[l,h] = v.slice();
+      return wide<U,N>(gather(p,l),gather(p,h));
+    }
+    // All other cases
+    else                                        return gather_(EVE_RETARGET(cpu_), p, v);
+  }
+}
+
+template<typename U, conditional_expr C, integral_scalar_value T, typename N>
+EVE_FORCEINLINE auto
+gather_(EVE_SUPPORTS(avx2_), C cx, U const* p, wide<T, N> v) noexcept -> wide<U, N>
+requires x86_abi<abi_t<T, N>>
+{
+  // Small index get converted then we recall gather
+  if constexpr(sizeof(T) < 4)
+  {
+    return gather[cx](p, convert(v, as<std::int32_t>{}));
+  }
+  else
+  {
+    if constexpr( C::is_complete) return gather(p, v);
+    else
+    {
+      using pl_t = long long const*;
+      using pi_t = int const*;
+
+      using enum category;
+      constexpr auto i = categorize<wide<T, N>>();
+      constexpr auto c = categorize<wide<U, N>>();
+
+      constexpr bool i_32x4   = match(i, int32x4 , uint32x4);
+      constexpr bool i_32x8   = match(i, int32x8 , uint32x8);
+      constexpr bool i_32x16  = match(i, int32x16, uint32x16);
+      constexpr bool i_64x2   = match(i, int64x2 , uint64x2);
+      constexpr bool i_64x4   = match(i, int64x4 , uint64x4);
+      constexpr bool i_64x8   = match(i, int64x8 , uint64x8);
+
+      constexpr bool d_i32x4  = match(c, int32x4 , uint32x4);
+      constexpr bool d_i32x8  = match(c, int32x8 , uint32x8);
+      constexpr bool d_i32x16 = match(c, int32x16, uint32x16);
+      constexpr bool d_i64x2  = match(c, int64x2 , uint64x2);
+      constexpr bool d_i64x4  = match(c, int64x4 , uint64x4);
+      constexpr bool d_i64x8  = match(c, int64x8 , uint64x8);
+
+      auto src = alternative(cx, wide<U, N>{}, as<wide<U, N>> {});
+      auto m = expand_mask(cx, as<wide<U, N>> {});
+
+      // 64 bits integer data
+      if      constexpr(i_32x8 && d_i64x8)  return _mm512_mask_i32gather_epi64(src,m.storage().value,v,p,8);
+      else if constexpr(i_32x4 && d_i64x2)  return _mm_mask_i32gather_epi64   (src,pl_t(p),v,m.mask(),8);
+      else if constexpr(i_32x4 && d_i64x4)  return _mm256_mask_i32gather_epi64(src,pl_t(p),v,m.mask(),8);
```

Comment:
CPUs supporting AVX512 would probably prefer `_mm_mmask` and `_mm256_mmask` over `_mm_mask` and `_mm256_mmask`, respectively.
I.e.,
```c++
_mm256_mmask_i32gather_epi64(src,m.storage().value,v,p,8);
```
But 128 bit and 256 bit vectors probably aren't a priority given AVX512, as these CPUs should probably be using 512 bit?

Reply:
Yeah, I somehow glazed over them!.
This will be better without need to go back to turning kmask into logicals.
I'm going to fix this tomorrow.

Reply:
fyi: limiting yourself to 32 byte registers on avx512 is very common bcausse the frequency scaling story on intel is atrocious.

Reply:
FWIW, I've never noticed frequency scaling being that much of an issue on Skylake-X.

Reply:
I personally only have SKX, CLX (SKX refresh), and TGL (laptop, 1 fma unit) hardware to test on. But SPR looks good: https://www.phoronix.com/review/intel-sapphirerapids-avx512

Reply:
Anyway, regardless of what benchmarks show, lots of people do like to use 256 bit vectors on AVX512.

Reply:
Seems like you know much more about it then I do. Anyways, we agree inn the end.

Reply:
Refactored the lot to have less if constexpr's